Go protos and better deps management

eyal-resonai opened the pull request:
Adding support for protos in go.

We link all go files source and all proto generated files into workspace, and all go files from dependencies.
We generate a go.mod file in the workspace to make it the root of the go project.
We create a go.mod file in proto dir with the same package name and add a 
```"replace proto=>./proto"``` directive in the go.mod.
We set the first dir in GOPATH to be yabtwork/go so that all downloaded packages are managed in the user machine and not inside the ephemeral docker
When we clean the workspace we make sure to keep the go.mod since if it is new go build redownload all packages (can we solve this?)

GoTest is doing exactly the same (with same function) just run go test instead of go build.
And if I got it right there is no need for GoLib since it will be identical to GoProg (didn't fully checked this theory yet)

eyal-resonai commented on tests/golang/YRoot:
```diff
@@ -1,3 +1,21 @@
+)
+
+DockerImage(
+    'builder-with-proto',
```
@itamaro , this image take quite some time to build since it build protobuf. Should we use the base image mechanism you are using in the proto test? can you push it to the same place?
I can create an image which both tests can use but since each test is a different project not sure how to do it in the current structure. WDYT?

itamaro replied:
easiest thing would be to move over this test under `tests/proto` so base image sharing is within the same project.

the other option would be creating a separate test project just for a common base image (that we'll build and push manually), and consume that image in both projects via `ExtDockerImage`.

is there a way to give you permissions to push to my docker hub repo?

eyal-resonai replied:
I've moved the test code from tests/golang to tests/proto and added the go and go protoc plugin into the image used there.
I've change the version and pushed it as ```yabt/proto-builder:3.5.1-go```.
yabt is a team I've created and added you to it.

eyal-resonai commented on yabt/target_utils_test.py:
```diff
@@ -93,7 +93,7 @@ def test_norm_name_unqualified_error():
 
 
-_HELLO_PROG_HASH = '0e049ee8ed31c616033076f8f90415a7'
+_HELLO_PROG_HASH = '19bd1e73278f9a524de049d0d950c3af'
```
I'm not sure why it changed. I'm guessing that since the proto builder signature have changed
But why didn't the proto traget hash changed?

itamaro replied:
> But why didn't the proto traget hash changed?

which proto target are you referring to? the one with the hash below this (`_PROTO_BUILDER`)? it doesn't have any dependencies on proto targets, so no reason for it to change.

eyal-resonai replied:
great

eyal-resonai commented:
@itamaro LMK if you have time and will to review. If not I'll press someone else :)

itamaro commented:
> @itamaro LMK if you have time and will to review. If not I'll press someone else :)

I should be able to review over the (long) weekend :-)

I noticed a lot of changes under `tests/golang` even after you moved to go+proto tests under `tests/proto` - can you explain what's going on?

maybe also rebase and cleanup the commit history? (I see this branch has my commits for Go builder that are already merged)

eyal-resonai commented:
I had bad conflicts when I've merged master into my branch I guess Shir who
started this PR branched of your branch and then you rebased it. Not sure
how easy it will be to squash it now. I'll give it a try.

I've moved only the go proto tests to proto but all the other go tests are
left on tests/golang (go images is small and I'm using a public images so
no issues in testing just simple go). Make sense?

On Wed, May 20, 2020, 18:34 Itamar Ostricher <notifications@github.com>
wrote:

> @itamaro <https://github.com/itamaro> LMK if you have time and will to
> review. If not I'll press someone else :)
>
> I should be able to review over the (long) weekend :-)
>
> I noticed a lot of changes under tests/golang even after you moved to
> go+proto tests under tests/proto - can you explain what's going on?
>
> maybe also rebase and cleanup the commit history? (I see this branch has
> my commits for Go builder that are already merged)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/resonai/ybt/pull/186#issuecomment-631551440>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJEMRJTPIJR6KPMALSAYDKTRSPZ6RANCNFSM4NFGQBSA>
> .
>

itamaro commented:
> I had bad conflicts when I've merged master into my branch I guess Shir who started this PR branched of your branch and then you rebased it. Not sure how easy it will be to squash it now. I'll give it a try.

merging master wouldn't work - but rebasing on top of master should. I didn't know Shir was building on top of my branch...


> I've moved only the go proto tests to proto but all the other go tests are left on tests/golang (go images is small and I'm using a public images so no issues in testing just simple go). Make sense?

it makes sense, the thing that confuses me is that i still see proto stuff under `tests/golang` - am I missing something?

eyal-resonai commented:
> > I had bad conflicts when I've merged master into my branch I guess Shir who started this PR branched of your branch and then you rebased it. Not sure how easy it will be to squash it now. I'll give it a try.
> 
> merging master wouldn't work - but rebasing on top of master should. I didn't know Shir was building on top of my branch...
I did the rebase. I'll squash after finishing code review?
> 
> > I've moved only the go proto tests to proto but all the other go tests are left on tests/golang (go images is small and I'm using a public images so no issues in testing just simple go). Make sense?
> 
> it makes sense, the thing that confuses me is that i still see proto stuff under `tests/golang` - am I missing something?
Sorry, it was dangling code I've removed it.

itamaro commented on yabt/builders/golang.py:
```diff
@@ -82,24 +84,140 @@ def go_prog_manipulate_target(build_context, target):
+    go_prog_builder_internal(build_context, target, command='test')
+
+
+def rm_all_but_go_mod(workspace_dir):
```
1. I don't like this
2. I don't understand the relation between a `go.mod` for a specific Prog/Test and the global `yabtwork/go` dir that stores downloaded packages. how does that work with multiple Go applications in the same project? by definition, to `go.mod` file will be different for each one, by they all use the same packages download dir? this feels risky and non-hermetic. what did we do about node_modules - it feels similar.

eyal-resonai commented on yabt/builders/golang.py:
```diff
@@ -82,24 +84,140 @@ def go_prog_manipulate_target(build_context, target):
     build_cmd_env.update(target.props.cmd_env or {})
+    build_cmd_env['GOPATH'] = ':'.join(gopaths)
+
+    if not isfile(join(workspace_dir, 'go.mod')):
```
done

eyal-resonai commented on yabt/builders/golang.py:
```diff
@@ -82,24 +84,140 @@ def go_prog_manipulate_target(build_context, target):
+              work_dir=join(buildenv_workspace, 'proto'))
+
+    bin_file = join(buildenv_workspace, binary)
+    build_cmd = ['go', command, '-o', bin_file] + buildenv_sources
```
this will be the executable which is the test (you can run it directly in any go env after it is built)

itamaro replied:
so running `go test -o foo ...` both builds a test binary and runs it?

eyal-resonai commented on yabt/builders/golang.py:
```diff
@@ -82,24 +84,140 @@ def go_prog_manipulate_target(build_context, target):
-        target.props.in_buildenv, build_cmd, build_cmd_env)
+      target.props.in_buildenv, build_cmd, build_cmd_env,
+      work_dir=buildenv_workspace)
     target.artifacts.add(
```
Morally, it is an artifact. We usually think of tests as terminal nodes in the build tree - but should we force it? 
This file is a movable executable which runs the tests (you can think of someone wants to have his own test driver which collect all the tests executable and run them on it's own).

eyal-resonai commented on yabt/builders/golang_test.py:
```diff
@@ -37,7 +37,39 @@ def test_golang_builder(basic_conf):
+ | | | | |  __/ | | | | | (_) |   | (_| | | |_) | | |_) |
+ |_| |_|  \___| |_| |_|  \___/     \__,_| | .__/  | .__/
+                                          |_|     |_|
+'''
```
fixed

eyal-resonai commented on yabt/builders/golang_test.py:
```diff
@@ -37,7 +37,39 @@ def test_golang_builder(basic_conf):
+ |_| |_|  \___| |_| |_|  \___/     \__,_| | .__/  | .__/
+                                          |_|     |_|
+'''
+    assert expected == (b"\n" + hello_out)
```
fixed

eyal-resonai commented on yabt/builders/proto.py:
```diff
@@ -47,10 +47,12 @@
      ('proto_cmd', PT.list, 'protoc'),
      ('gen_python', PT.bool, True),
      ('gen_cpp', PT.bool, True),
+     ('gen_go', PT.bool, False),
```
this will force all the users of ybt to update all images used to run protoc to have the go plugin in them.
Currently we are forcing cpp and cpp but I think both of them comes with the default installation of protobuf lib while go doesn't

eyal-resonai commented on yabt/builders/proto.py:
```diff
@@ -47,10 +47,12 @@
      ('gen_cpp_rpcz', PT.bool, False),
      ('gen_python_grpc', PT.bool, False),
      ('gen_cpp_grpc', PT.bool, False),
+     ('gen_go_grpc', PT.bool, False),
```
manually (we already using that in resonai code base)

eyal-resonai commented on yabt/config.py:
```diff
@@ -121,6 +121,10 @@ def get_build_file_path(self, build_module) -> str:
 
+    def get_root_workspace_path(self) -> str:
+        return os.path.join(
+            self.project_root, self.builders_workspace_dir)
```
done

itamaro commented on tests/golang/hello/YBuild:
```diff
@@ -1,3 +1,12 @@
+
+GoTest('hello_test',
+       sources='hello_test.go',
+       deps=':hello',
```
It's not about testing functions in main module, it's about forcing the Prog to produce the binary artifact when you just want to test - this could be expensive.

Is there a way around it?

itamaro commented on tests/golang/hello/go.mod:
```diff
@@ -0,0 +1,6 @@
+module foo.com
```
1. the thing I'm not sure I understand here is, with `yabtwork/go` being shared by all Go targets, but different targets having different `go.mod` files, and with the observation that `go.mod` "changes" cause go to redownload the packages - doesn't that mean that once you have at least 2 different `go.mod` files in the same project, you're guaranteed to have all packages redownloaded every build?
2. it sounds like `go.mod` is similar to Python `requirements.txt` files, which ybt has adopted a philosophy of generating based on `PythonPackage` targets. Shouldn't we strive for consistent philosophy? (it's ok if we're doing a shortcut to get it working, but I'd like to understand what is the desired longterm situation)

itamaro commented on tests/proto/3rdparty/YBuild:
```diff
@@ -24,3 +24,17 @@ PythonPackage(
+    script='install-golang.sh',
+    packaging_params={
+        'set_env': {
+              'GOPATH': '/go',
+         },
+        'semicolon_join_env': {
+            'PATH': '/usr/local/go/bin:/go/bin',
+        },
```
something is still off with the indentation here

itamaro commented on tests/proto/3rdparty/install-golang.sh:
```diff
@@ -0,0 +1,14 @@
+mkdir "$GOPATH"
+# Support for grpc generation.
+apt update
+apt-get install -y git
```
leave TODO to clean it up?

itamaro commented on yabt/builders/golang.py:
```diff
@@ -82,24 +84,140 @@ def go_prog_manipulate_target(build_context, target):
+    builder_name = target.builder_name
+    yprint(build_context.conf, 'Build', builder_name, target)
+    workspace_dir = build_context.get_workspace(builder_name, target.name)
+    go_package = (target.props.get('go_package') or
+                  build_context.conf.get('go_package', None))
```
hm, surprising

itamaro commented on yabt/artifact.py:
```diff
@@ -70,6 +71,10 @@ class ArtifactStore:
                 by any dependent C++ builder (direct or indirect) to extend
                 the target headers. Cached as ouputs.
+    - "gen_go"  A generated Go source file (e.g. from Proto's) that is used
+                only by directly dependent Go builders (used to extend
```
given the comment about wanting transitive deps too, change the comment here to reflect that
```suggestion
                by any dependent Go builders (direct or indirect) (used to extend
```

itamaro commented on yabt/builders/golang.py:
```diff
@@ -82,24 +84,140 @@ def go_prog_manipulate_target(build_context, target):
+    sources_to_link = list(target.props.sources)
+    has_protos = False
+    for dep in build_context.generate_all_deps(target):
+        sources_to_link.extend(dep.props.get('sources', []))
```
probably doesn't matter for now. I guess if at some point we support C-Go bindings, this could get confusing and cluttered.
maybe leave TODO?

itamaro commented on yabt/builders/golang.py:
```diff
@@ -82,24 +84,140 @@ def go_prog_manipulate_target(build_context, target):
+      join(build_context.conf.get_root_workspace_path(), 'go'))
+
+    gopaths = [download_cache_dir]
+    user_gopath = (target.props.cmd_env or {}).get('GOPATH')
```
oh, I see `cmd_env` prop is not typed as dict - not sure what's the reason, but I guess it's fine.

itamaro commented on yabt/config.py:
```diff
@@ -121,6 +121,10 @@ def get_build_file_path(self, build_module) -> str:
 
+    def get_go_packages_path(self) -> str:
+        return os.path.join(
+            self.project_root, self.builders_workspace_dir, 'go')
```
for cache we use `.cache`, so maybe `.go` for consistency? don't feel strongly about this.
